usso.go

```diff
@@ -95,6 +98,47 @@ func (server UbuntuSSOServer) GetTokenWithOTP(email, password, otp, tokenName st
 	return &ssodata, nil
 }
 
+// Error represents an error message returned from Ubuntu SSO.
+type Error struct {
+	Message string                 `json:"message"`
+	Code    string                 `json:"code,omitempty"`
+	Extra   map[string]interface{} `json:"extra,omitempty"`
+}
+
+// getError attempts to extract the most meaningfule error that it can
+// from a response.
+func getError(resp *http.Response) *Error {
+	var ssoError Error
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		ssoError.Code = resp.Status
+		ssoError.Message = resp.Status
+		return &ssoError
+	}
+	err = json.Unmarshal(body, &ssoError)
+	if err != nil {
+		// Attempt to pass the original error back in the best way possible
+		ssoError.Code = resp.Status
+		ssoError.Message = string(body)
+		return &ssoError
+	}
+	return &ssoError
+}
+
+// Error implements error.Error.
+func (err *Error) Error() string {
+	if len(err.Extra) == 0 {
+		return err.Message
+	}
+	extra := new(bytes.Buffer)
+	for k, v := range err.Extra {
+		fmt.Fprintf(extra, "%s: %v, ", k, v)
+	}
+	extra.Truncate(extra.Len() - 2)
```

Comment:
it's unlikely, but meddling in the above lines could cause this to panic one day. Might we want to protect ourselves?